testcases/kernel/syscalls/fsetxattr/fsetxattr02.c

```diff
@@ -0,0 +1,267 @@
+#define BLK      "fsetxattr02blk"
+#define SOCK     "fsetxattr02sock"
+
+struct test_case {
+	enum test_type {
+		TP_FILE,
+		TP_DIRECTORY,
+		TP_SYMLINK,
+		TP_FIFO,
+		TP_CHARDEV,
+		TP_BLOCKDEV,
+		TP_SOCKET
+	} type;
+	char *fname;
+	int filed;
+	int fflags;
+	char *key;
+	char *value;
+	size_t size;
+	int flags;
+	int exp_err;
+};
+static struct test_case tc[] = {
+	{			/* case 00, set attr to reg */
+	 .type = TP_FILE,
+	 .fname = FILENAME,
+	 .filed = -1,
+	 .fflags = O_RDONLY,
+	 .key = XATTR_TEST_KEY,
+	 .value = XATTR_TEST_VALUE,
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = 0,
+	 },
+	{			/* case 01, set attr to dir */
+	 .type = TP_DIRECTORY,
+	 .fname = DIRNAME,
+	 .filed = -1,
+	 .fflags = O_RDONLY,
+	 .key = XATTR_TEST_KEY,
+	 .value = XATTR_TEST_VALUE,
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = 0,
+	 },
+	{			/* case 02, set attr to symlink */
+	 .type = TP_SYMLINK,
+	 .fname = SYMLINK,
+	 .filed = -1,
+	 .fflags = O_RDONLY,
+	 .key = XATTR_TEST_KEY,
+	 .value = XATTR_TEST_VALUE,
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = EEXIST,
+	 },
+	{			/* case 03, set attr to fifo */
+	 .type = TP_FIFO,
+	 .fname = FIFO,
+	 .filed = -1,
+	 .fflags = (O_RDONLY | O_NONBLOCK),
+	 .key = XATTR_TEST_KEY,
+	 .value = XATTR_TEST_VALUE,
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = EPERM,
+	 },
+	{			/* case 04, set attr to character special */
+	 .type = TP_CHARDEV,
+	 .fname = CHR,
+	 .filed = -1,
+	 .fflags = O_RDONLY,
+	 .key = XATTR_TEST_KEY,
+	 .value = XATTR_TEST_VALUE,
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = EPERM,
+	 },
+	{			/* case 05, set attr to block special */
+	 .type = TP_BLOCKDEV,
+	 .fname = BLK,
+	 .filed = -1,
+	 .fflags = O_RDONLY,
+	 .key = XATTR_TEST_KEY,
+	 .value = XATTR_TEST_VALUE,
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = EPERM,
+	 },
+	{			/* case 06, set attr to socket */
+	 .type = TP_SOCKET,
+	 .fname = SOCK,
+	 .filed = -1,
+	 .fflags = O_RDONLY,
+	 .key = XATTR_TEST_KEY,
+	 .value = XATTR_TEST_VALUE,
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = EPERM,
+	 },
+};
+
+static void verify_fsetxattr(unsigned int i)
+{
+	TEST(fsetxattr(tc[i].filed, tc[i].key, tc[i].value, tc[i].size,
+			tc[i].flags));
+
+	if (TEST_RETURN == -1 && TEST_ERRNO == EOPNOTSUPP)
+		tst_brk(TCONF, "fsetxattr() not supported");
+
+	if (!tc[i].exp_err) {
+		if (TEST_RETURN) {
+			tst_res(TFAIL | TTERRNO,
+				"fsetxattr() failed with %li", TEST_RETURN);
+			return;
+		}
+
+		tst_res(TPASS, "fsetxattr() passed");
+		return;
+	}
+
+	if (TEST_RETURN == 0) {
+		tst_res(TFAIL, "fsetxattr() passed unexpectedly");
+		return;
+	}
+
+	if (TEST_ERRNO != tc[i].exp_err) {
+		tst_res(TFAIL | TTERRNO, "fsetxattr() should fail with %s",
+			tst_strerrno(tc[i].exp_err));
+		return;
+	}
+
+	tst_res(TPASS | TTERRNO, "fsetxattr() failed");
+}
```

Comment:
It would probably be a bit nicer if we included what type of fd we were testing, i.e. use the fname in the tst_res() messages so that we have something as:

```
fsetxattr02.c:87: PASS: fsetxattr() on testdir passed
fsetxattr02.c:87: PASS: fsetxattr() on symlink failed: EEXIST
````

rather than:

```
fsetxattr02.c:87: PASS: fsetxattr() passed
fsetxattr02.c:87: PASS: fsetxattr() failed: EEXIST
```

Reply:
Will do.